language/diem-vm/src/diem_transaction_executor.rs

```diff
@@ -172,16 +172,30 @@ impl DiemVM {
             cost_strategy
                 .charge_intrinsic_gas(txn_data.transaction_size())
                 .map_err(|e| e.into_vm_status())?;
-            session
-                .execute_script(
+
+            match payload {
+                TransactionPayload::Script(script) => session.execute_script(
                     script.code().to_vec(),
                     script.ty_args().to_vec(),
                     convert_txn_args(script.args()),
                     vec![txn_data.sender()],
                     cost_strategy,
                     log_context,
-                )
-                .map_err(|e| e.into_vm_status())?;
+                ),
+                TransactionPayload::ScriptFunction(script_fn) => session.execute_script_function(
```

Comment:
Also this will be a breaking change in code behavior: the old Diem validator won't understand this feature until the new binary gets deployed. Thus we need to make sure this feature is gated in the DiemVM.